segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3Storage.java

```diff
@@ -0,0 +1,380 @@
+        String uploadId = client.initiateMultipartUpload(config.getBucket(), targetPath);
+
+        // check whether the target exists
+        if (!syncExists(targetHandle.getSegmentName())) {
+            throw new StreamSegmentNotExistsException(targetHandle.getSegmentName());
+        }
+        // check whether the source is sealed
+        SegmentProperties si = syncGetStreamSegmentInfo(sourceSegment);
+        Preconditions.checkState(si.isSealed(), "Cannot concat segment '%s' into '%s' because it is not sealed.", sourceSegment, targetHandle.getSegmentName());
+
+        //Upload the first part
+        CopyPartRequest copyRequest = new CopyPartRequest(config.getBucket(),
+                targetPath,
+                config.getBucket(),
+                targetPath,
+                uploadId,
+                1).withSourceRange(Range.fromOffsetLength(0, offset));
+        CopyPartResult copyResult = client.copyPart(copyRequest);
+
+        partEtags.add(new MultipartPartETag(copyResult.getPartNumber(), copyResult.getETag()));
+
+        //Upload the second part
+        S3ObjectMetadata metadataResult = client.getObjectMetadata(config.getBucket(),
+                config.getRoot() + sourceSegment);
+        long objectSize = metadataResult.getContentLength(); // in bytes
+
+        copyRequest = new CopyPartRequest(config.getBucket(),
+                config.getRoot() + sourceSegment,
+                config.getBucket(),
+                targetPath,
+                uploadId,
+                2).withSourceRange(Range.fromOffsetLength(0, objectSize));
+
+        copyResult = client.copyPart(copyRequest);
+        partEtags.add(new MultipartPartETag(copyResult.getPartNumber(), copyResult.getETag()));
+
+        //Close the upload
+        client.completeMultipartUpload(new CompleteMultipartUploadRequest(config.getBucket(),
+                targetPath, uploadId).withParts(partEtags));
+
+        client.deleteObject(config.getBucket(), config.getRoot() + sourceSegment);
+        LoggerHelpers.traceLeave(log, "concat", traceId);
+
+        return null;
+    }
+
+    private Void syncDelete(SegmentHandle handle) {
+
+        client.deleteObject(config.getBucket(), config.getRoot() + handle.getSegmentName());
+        return null;
+    }
+
+    @Override
+    protected Throwable translateException(String segmentName, Throwable e) {
+        Throwable retVal = e;
+
+        if (e instanceof S3Exception && !Strings.isNullOrEmpty(((S3Exception) e).getErrorCode())) {
+            if (((S3Exception) e).getErrorCode().equals("NoSuchKey")) {
+                retVal = new StreamSegmentNotExistsException(segmentName);
+            }
+            if (((S3Exception) e).getErrorCode().equals("InvalidRange")) {
+                retVal = new IllegalArgumentException(segmentName, e);
+            }
+        }
+
+        if (e instanceof IndexOutOfBoundsException) {
+            retVal = new ArrayIndexOutOfBoundsException(e.getMessage());
```

Comment:
The Storage contract expects `ArrayIndexOutOfBoundsException`. This is just translating from `IndexOutOfBounds` to `ArrayIndexOutOfBoundsException`. Earlier code was also translating other exceptions to `IllegalArgumentException`.